pyiceberg/table/update/snapshot.py

```diff
@@ -477,6 +478,20 @@ def _deleted_entries(self) -> List[ManifestEntry]:
         return []
 
 
+@dataclass(init=False)
```

Comment:
What do you think of making this `Frozen`? This gives some nice benefits like being hashable: https://docs.python.org/3/library/dataclasses.html#frozen-instances

Using the `default_factory'`s, we can also drop the init.

```suggestion
@dataclass(frozen=True)
```